### PR TITLE
MUDROD 132

### PR DESCRIPTION
**Pull request by Yongyao**
Commands cleanup. A documentation will be posted here later on

### CODE REVIEW

**Comment by lewismc** [comment]
@Yongyao can you or @quintinali please get a page together demonstrating the workflow? Can you base it off of using the PR you've provided here and using the Logs staged by Chris finch?
Thanks

**Comment by lewismc** [comment]
Also can we use tags and milestones for tracking all of this stuff? It makes it MUCH easier to review when we have our meetings. Thanks.

**Comment by Yongyao** [comment]
https://github.com/mudrod/mudrod/wiki/Command-instructions
@lewismc @fgreg @quintinali Here is a detailed documentation. All of us should test it as this is very important!!!

**Comment by lewismc** [comment]
Hi @Yongyao this is great, but it doesn't document the workflow. Which flags does someone need to use, in which order to achieve full log ingestion for Mudrod? Thats what we need to address here. Thanks.

**Comment by Yongyao** [comment]
@lewismc workflow added. https://github.com/mudrod/mudrod/wiki/Command-Line-Interface

**Comment by fgreg** [comment]
When using `-f` what should `-dataDir` be set to? What should the structure of that directory look like? Should it be something like:

```
dataDir/
    /logs
        /FTP.201611.gz
        /LAS.201611.gz
        /OPeNDAP.201611.gz
        /THREDDS.201611.gz
        /WWW.201611.gz
    /RawMetadata
```

If `-f` needs both logs and metadata then there should be an option to configure the location of both.


FYI - This is the directory structure I am working with for the logs:

```
mudrod-ingest/
├── ftp-mirror
│   ├── 2016
│   │   ├── 11
│   │   │   ├── FTP.201611.gz
│   │   │   ├── LAS.201611.gz
│   │   │   ├── OPeNDAP.201611.gz
│   │   │   ├── THREDDS.201611.gz
│   │   │   └── WWW.201611.gz
│   │   └── 12
│   │       ├── EMS_LAS.201612.gz
│   │       ├── EMS_OPeNDAP.201612.gz
│   │       ├── EMS_THREDDS.201612.gz
│   │       └── README
│   └── 2017
│       └── 01
│           ├── FTP.201701.gz
│           ├── LAS.201701.gz
│           ├── OPeNDAP.201701.gz
│           ├── THREDDS.201701.gz
│           └── WWW.201701.gz
```

**Comment by Yongyao** [comment]
When using -f, the directory structure should look like below
```
dataDir/
    /httplog
        /httplog
    /ftplog
        /ftplog
    /RawMetadata
        /1.json
        /2.json
```
This code can find those log files in this structure as long as you set the prefix correct.

**Comment by lewismc** [comment]
We need to augment the workflow with this information. The pattern of the logs should mimic the formatting given to us by Chris Finch.

**Comment by Yongyao** [comment]
Did everyone get a chance to test this?

**Comment by fgreg** [comment]
I have not. Quick comment about the documentation though.

" You can change it from here: https://github.com/mudrod/mudrod/blob/master/core/src/main/resources/config.xml#L47" changing the config.xml file when you are trying to run the command-line app will accomplish nothing at this point since the artifact is already built.

However I did add the `MUDROD_CONFIG` environment variable which can specify the location to a config.xml file that will be used in place of the one packaged with the application.

So the instructions should read something along the lines of: copy config.xml to the same directory you are running mudrod-engine from; modify that the copied config.xml file; set the `MUDROD_CONFIG` environment variable to the absolute path of your config; run mudrod-engine.

Also add in the information about expected directory structure.

File: core/src/main/java/gov/nasa/jpl/mudrod/discoveryengine/RecommendEngine.java [hunk 1]
```diff
@@ -21,13 +21,11 @@
...
   public RecommendEngine(Properties props, ESDriver es, SparkDriver spark) {
     super(props, es, spark);
-    // TODO Auto-generated constructor stub
     LOG.info("Started Mudrod Recommend Engine.");
   }
 
   @Override
   public void preprocess() {
-    // TODO Auto-generated method stub
     LOG.info("*****************Recommendation preprocessing starts******************");
```
Comment (lewismc) [inline core/src/main/java/gov/nasa/jpl/mudrod/discoveryengine/RecommendEngine.java:29]: We really ought to get rid of this logging. It is totally unnecessary. All we need is a simple log message and possibly a timing metric so we can determine, at a glance, how long each stage took to complete or fail.

File: core/src/main/java/gov/nasa/jpl/mudrod/main/MudrodEngine.java [hunk 2]
```diff
@@ -212,70 +210,44 @@ private String decompressSVMWithSGDModel(String archiveName) throws IOException
...
+  public void startLogIngest() {
     DiscoveryEngineAbstract wd = new WeblogDiscoveryEngine(props, es, spark);
     wd.preprocess();
     wd.process();
-
-    DiscoveryEngineAbstract od = new OntologyDiscoveryEngine(props, es, spark);
-    od.preprocess();
-    od.process();
-
+    LOG.info("*****************logs have been ingested successfully******************");
```
Comment (lewismc) [inline core/src/main/java/gov/nasa/jpl/mudrod/main/MudrodEngine.java:221]: Same with this logging... there is no need for this.

File: core/src/main/java/gov/nasa/jpl/mudrod/main/MudrodEngine.java [hunk 3]
```diff
@@ -212,70 +210,44 @@ private String decompressSVMWithSGDModel(String archiveName) throws IOException
...
+  }
+  
+  /**
+   * updating and analysing metadata to metadata similarity results
```
Comment (lewismc) [inline core/src/main/java/gov/nasa/jpl/mudrod/main/MudrodEngine.java:225]: This comment doesn't make any sense at all.

File: core/src/main/java/gov/nasa/jpl/mudrod/main/MudrodEngine.java [hunk 4]
```diff
@@ -333,24 +296,19 @@ public static void main(String[] args) {
...
-    // calculate vocab similarity from session reconstrution results
-    Option vocabSimFromOpt = new Option("v", VOCAB_SIM_FROM_LOG, false, "begin similarity calulation from web log Mudrod workflow");
-    // add metadata and ontology preprocessing and processing results into web
-    // log vocab similarity
-    Option addMetaOntoOpt = new Option("a", ADD_META_ONTO, false, "begin adding metadata and ontology results");
 
     // argument options
-    Option logDirOpt = OptionBuilder.hasArg(true).withArgName("/path/to/log/directory").hasArgs(1).withDescription("the log directory to be processed by Mudrod").withLongOpt("logDirectory")
-        .isRequired().create(LOG_DIR);
+    Option dataDirOpt = OptionBuilder.hasArg(true).withArgName("/path/to/data/directory").hasArgs(1).withDescription("the data directory to be processed by Mudrod").withLongOpt("logDirectory")
```
Comment (lewismc) [inline core/src/main/java/gov/nasa/jpl/mudrod/main/MudrodEngine.java:310]: The long description should change if the short description is changing.
```.withLongOpt("logDirectory")``` should be ```.withLongOpt("dataDirectory")```

**Comment by lewismc** [comment]
I've just corrected a bunch of the Command Line Documentation folks as they are incorrect...

**Comment by lewismc** [comment]
When applying this patch, building against master and attempting to ingest logs I get the following
```
lmcgibbn@LMC-056430 /usr/local/mudrod(MUDROD-132) $ ./core/target/appassembler/bin/mudrod-engine -f -dataDir /usr/local/mudrod_resources/2016/11/
2017-05-03 01:39:05,102 INFO  main.MudrodEngine (MudrodEngine.java:locateConfig(124)) - Loaded config file from /usr/local/mudrod/core/src/main/resources/config.xml
2017-05-03 01:39:06,684 WARN  util.NativeCodeLoader (NativeCodeLoader.java:<clinit>(62)) - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2017-05-03 01:39:06,748 WARN  util.Utils (Logging.scala:logWarning(66)) - Your hostname, LMC-056430 resolves to a loopback address: 127.0.0.1; using 192.168.1.76 instead (on interface en0)
2017-05-03 01:39:06,748 WARN  util.Utils (Logging.scala:logWarning(66)) - Set SPARK_LOCAL_IP if you need to bind to another address
2017-05-03 01:39:07,082 INFO  util.log (Log.java:initialized(186)) - Logging initialized @2279ms
2017-05-03 01:39:07,170 INFO  server.Server (Server.java:doStart(327)) - jetty-9.2.z-SNAPSHOT
2017-05-03 01:39:07,187 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@3e05586b{/jobs,null,AVAILABLE}
2017-05-03 01:39:07,187 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@35b17c06{/jobs/json,null,AVAILABLE}
2017-05-03 01:39:07,188 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@45545e7a{/jobs/job,null,AVAILABLE}
2017-05-03 01:39:07,188 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@9f674ac{/jobs/job/json,null,AVAILABLE}
2017-05-03 01:39:07,188 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@1da4b3f9{/stages,null,AVAILABLE}
2017-05-03 01:39:07,189 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@23cbbd07{/stages/json,null,AVAILABLE}
2017-05-03 01:39:07,189 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@448b808a{/stages/stage,null,AVAILABLE}
2017-05-03 01:39:07,189 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@7e62cfa3{/stages/stage/json,null,AVAILABLE}
2017-05-03 01:39:07,190 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@46e190ed{/stages/pool,null,AVAILABLE}
2017-05-03 01:39:07,190 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@383c3eb3{/stages/pool/json,null,AVAILABLE}
2017-05-03 01:39:07,192 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@20ab76ee{/storage,null,AVAILABLE}
2017-05-03 01:39:07,192 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@485caa8f{/storage/json,null,AVAILABLE}
2017-05-03 01:39:07,192 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@2703d91{/storage/rdd,null,AVAILABLE}
2017-05-03 01:39:07,193 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@5be052ca{/storage/rdd/json,null,AVAILABLE}
2017-05-03 01:39:07,193 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@2007435e{/environment,null,AVAILABLE}
2017-05-03 01:39:07,194 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@da28d03{/environment/json,null,AVAILABLE}
2017-05-03 01:39:07,194 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@ebda593{/executors,null,AVAILABLE}
2017-05-03 01:39:07,194 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@5792c08c{/executors/json,null,AVAILABLE}
2017-05-03 01:39:07,195 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@4d157493{/executors/threadDump,null,AVAILABLE}
2017-05-03 01:39:07,195 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@54c622a7{/executors/threadDump/json,null,AVAILABLE}
2017-05-03 01:39:07,201 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@37348491{/static,null,AVAILABLE}
2017-05-03 01:39:07,202 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@76e3b45b{/,null,AVAILABLE}
2017-05-03 01:39:07,203 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@10b4e7f8{/api,null,AVAILABLE}
2017-05-03 01:39:07,203 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@75023c53{/jobs/job/kill,null,AVAILABLE}
2017-05-03 01:39:07,204 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@66813e6e{/stages/stage/kill,null,AVAILABLE}
2017-05-03 01:39:07,209 INFO  server.ServerConnector (AbstractConnector.java:doStart(266)) - Started ServerConnector@1ddc6db2{HTTP/1.1}{0.0.0.0:4040}
2017-05-03 01:39:07,210 INFO  server.Server (Server.java:doStart(379)) - Started @2408ms
2017-05-03 01:39:07,470 INFO  handler.ContextHandler (ContextHandler.java:doStart(744)) - Started o.s.j.s.ServletContextHandler@626d2016{/metrics/json,null,AVAILABLE}
2017-05-03 01:39:07,618 INFO  discoveryengine.WeblogDiscoveryEngine (WeblogDiscoveryEngine.java:<init>(51)) - Started Mudrod Weblog Discovery Engine.
2017-05-03 01:39:07,618 INFO  discoveryengine.WeblogDiscoveryEngine (WeblogDiscoveryEngine.java:preprocess(98)) - Starting Web log preprocessing.
usage: MudrodEngine: 'logDir' argument is mandatory. User must also
       provide an ingest method. -dataDir </path/to/data/directory>
       [-esHost <host_name>] [-esPort <port_num>] [-esTCPPort <port_num>]
       [-f] [-h] [-l] [-m] [-p]
 -dataDir,--logDirectory </path/to/data/directory>       the data
                                                         directory to be
                                                         processed by
                                                         Mudrod
 -esHost,--elasticSearchHost <host_name>                 elasticsearch
                                                         cluster unicast
                                                         host
 -esPort,--elasticSearchHTTPPort <port_num>              elasticsearch
                                                         HTTP/REST port
 -esTCPPort,--elasticSearchTransportTCPPort <port_num>   elasticsearch
                                                         transport TCP
                                                         port
 -f,--fullIngest                                         begin full ingest
                                                         Mudrod workflow
 -h,--help                                               show this help
                                                         message
 -l,--logIngest                                          begin log ingest
 -m,--metaIngest                                         begin metadata
                                                         ingest
 -p,--processingWithPreResults                           begin processing
                                                         with
                                                         preprocessing
                                                         results
2017-05-03 01:39:07,620 ERROR main.MudrodEngine (MudrodEngine.java:main(399)) - Error whilst parsing command line.
java.lang.NullPointerException
	at gov.nasa.jpl.mudrod.discoveryengine.WeblogDiscoveryEngine.getFileList(WeblogDiscoveryEngine.java:63)
	at gov.nasa.jpl.mudrod.discoveryengine.WeblogDiscoveryEngine.preprocess(WeblogDiscoveryEngine.java:100)
	at gov.nasa.jpl.mudrod.main.MudrodEngine.startFullIngest(MudrodEngine.java:240)
	at gov.nasa.jpl.mudrod.main.MudrodEngine.main(MudrodEngine.java:389)
2017-05-03 01:39:07,626 INFO  server.ServerConnector (AbstractConnector.java:doStop(306)) - Stopped ServerConnector@1ddc6db2{HTTP/1.1}{0.0.0.0:4040}
2017-05-03 01:39:07,627 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@66813e6e{/stages/stage/kill,null,UNAVAILABLE}
2017-05-03 01:39:07,628 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@75023c53{/jobs/job/kill,null,UNAVAILABLE}
2017-05-03 01:39:07,628 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@10b4e7f8{/api,null,UNAVAILABLE}
2017-05-03 01:39:07,628 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@76e3b45b{/,null,UNAVAILABLE}
2017-05-03 01:39:07,628 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@37348491{/static,null,UNAVAILABLE}
2017-05-03 01:39:07,628 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@54c622a7{/executors/threadDump/json,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@4d157493{/executors/threadDump,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@5792c08c{/executors/json,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@ebda593{/executors,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@da28d03{/environment/json,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@2007435e{/environment,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@5be052ca{/storage/rdd/json,null,UNAVAILABLE}
2017-05-03 01:39:07,629 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@2703d91{/storage/rdd,null,UNAVAILABLE}
2017-05-03 01:39:07,630 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@485caa8f{/storage/json,null,UNAVAILABLE}
2017-05-03 01:39:07,630 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@20ab76ee{/storage,null,UNAVAILABLE}
2017-05-03 01:39:07,630 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@383c3eb3{/stages/pool/json,null,UNAVAILABLE}
2017-05-03 01:39:07,630 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@46e190ed{/stages/pool,null,UNAVAILABLE}
2017-05-03 01:39:07,630 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@7e62cfa3{/stages/stage/json,null,UNAVAILABLE}
2017-05-03 01:39:07,630 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@448b808a{/stages/stage,null,UNAVAILABLE}
2017-05-03 01:39:07,631 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@23cbbd07{/stages/json,null,UNAVAILABLE}
2017-05-03 01:39:07,631 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@1da4b3f9{/stages,null,UNAVAILABLE}
2017-05-03 01:39:07,631 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@9f674ac{/jobs/job/json,null,UNAVAILABLE}
2017-05-03 01:39:07,631 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@45545e7a{/jobs/job,null,UNAVAILABLE}
2017-05-03 01:39:07,631 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@35b17c06{/jobs/json,null,UNAVAILABLE}
2017-05-03 01:39:07,631 INFO  handler.ContextHandler (ContextHandler.java:doStop(865)) - Stopped o.s.j.s.ServletContextHandler@3e05586b{/jobs,null,UNAVAILABLE}
```

**Comment by lewismc** [comment]
Right now I am struggling to ingest anything into ES... not sure how and why this got so complicated. We need to simplify it significantly... right now it's a bit of a mess to be honest and it's wasting time.

**Comment by Yongyao** [comment]
@lewismc I think this error is due to the incorrect data directory structure as @fgreg and I discussed earlier in this thread. Could you please tell me what logs you were using? I saw that you emailed Chris that you cannot access those logs anymore.

In short, to solve this problem, we just need to reach an agreement on what data directory structure,  file format we should be using.

**Comment by lewismc** [comment]
Thanks @Yongyao, the directory structure is as follows
```
lmcgibbn@LMC-056430 /usr/local/mudrod_resources/2016/11 $ tree
.
├── FTP.201611
├── LAS.201611
├── OPeNDAP.201611
├── THREDDS.201611
├── UserHistoryMatrix.csv
└── WWW.201611

0 directories, 6 files
```
The **UserHistoryMatrix.csv** was generated during the full log ingestion using master branch. Everything else was provided directly by Chris Finch and we should make interpretation of these files default. Anything else will not be used within MUDROD anymore.

I'm running another log ingestion using master branch (the PR associated with this issue did not work) so I will update once this completes.

**Comment by fgreg** [comment]
@lewismc @Yongyao I've fixed one of the errors but have run into a new one. The problem is that this pull request was created from Yongyao's fork so I can't add my commits to this pull request. I have now created a [new branch](https://github.com/mudrod/mudrod/tree/mudrod-132) that incorporates both Yongyao's changes and mine.

I suggest we close this pull request and open a new one from the feature branch https://github.com/mudrod/mudrod/tree/mudrod-132 so that we are all able to contribute commits.

**Comment by Yongyao** [comment]
OK, please feel free to do that. Also, I still cannot access the ftp page Chris set up. Could you please share the data you are using so that I can help debug this issue too?

**Comment by fgreg** [comment]
This discussion continued in PR https://github.com/mudrod/mudrod/pull/159

**Pull request closed without merging**